runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformMatchers.java

```diff
@@ -158,4 +160,39 @@ public boolean matches(AppliedPTransform<?, ?, ?> application) {
       }
     };
   }
+
+  /**
+   * A {@link PTransformMatcher} which matches a {@link ParDo.Bound} or {@link ParDo.BoundMulti}
+   * where the {@link DoFn} is of the provided type.
+   */
+  public static PTransformMatcher parDoWithFnType(final Class<? extends DoFn> fnType) {
+    return new PTransformMatcher() {
+      @Override
+      public boolean matches(AppliedPTransform<?, ?, ?> application) {
+        DoFn<?, ?> fn;
+        if (application.getTransform() instanceof ParDo.Bound) {
+          fn = ((ParDo.Bound) application.getTransform()).getFn();
```

Comment:
This code basically does this, but I just want to make sure that your intent matches how it should work: In a world where this matcher is running over a reconstructed graph, the matcher just fails to match anything that isn't deserialized into known Java-SDK-specific classes, and also fails to match if the DoFn is junk from another language.